cunumeric/array.py

```diff
@@ -2464,6 +2456,50 @@ def diagonal(
                 raise ValueError("Either axis1/axis2 or axes must be supplied")
         return self._diag_helper(offset=offset, axes=axes, extract=extract)
 
+    @add_boilerplate("indices", "values")
+    def put(
+        self, indices: ndarray, values: ndarray, mode: str = "raise"
+    ) -> None:
+        """
+        Replaces specified elements of the array with given values.
+
+        Refer to :func:`cunumeric.put` for full documentation.
+
+        See Also
+        --------
+        cunumeric.put : equivalent function
+
+        Availability
+        --------
+        Multiple GPUs, Multiple CPUs
+
+        """
+
+        if values.size == 0 or indices.size == 0:
+            return
+
+        if mode not in ("raise", "wrap", "clip"):
+            raise ValueError(
+                "mode must be one of 'clip', 'raise', or 'wrap' "
+                f"(got  {mode})"
+            )
+
+        if mode == "wrap":
+            indices = indices % self.size
+        elif mode == "clip":
+            indices = indices.clip(0, self.size - 1)
```

Comment:
I don't think we currently do any checks in the case where `mode == "raise"`.

Reply:
that is right, just pushed the check for CPU and OpenMP. To my understanding we are not allowed to throw exceptions in CUDA kernels, is that correct?

Reply:
Correct, you would need to have the kernels signal this failure in a different way, e.g. reducing into a shared boolean variable that an out-of-bounds index was found. Or run a second kernel that looks for such indices.

But at that point it may be cleaner to not do this check inside the wrap task, instead add a 3rd case to `array.py`:
```
else:  # mode == "raise"
  if indices.min() < 0 or indices.max() >= self.size:
    raise ...
```
This way you also skip the check in case the user passes "wrap" or "clip" (the check is unnecessary in that case).

Reply:
Now that you mention it, wouldn't this also apply to the zip kernel? Isn't it also bad that `compute_idx` throws an exception there as well?

Reply:
That makes sense. 
However, the down side of having the check on the python side is that it will invoke 2 more tasks: `min` and `max`. So,  I think it would be better to make CUDA task to return error . What do you think?
And yes, the same issue applies to the ZIP kernel. In that case compiler didn't complain about exception because it is hidden in the constexpr function

Reply:
Yes, good point. The simplest solution might be to run a second kernel (in the wrap and zip tasks) to look for out-of-bounds indices - `thrust::minmax_element` might be a good choice. And then don't do the bounds checks in the "main" wrap/zip kernel (you would need to have a separate `compute_idx` function I assume for .cu files).